### DISCUSSION TITLE
Button to process collected shaders

### DISCUSSION THREAD

**Discussion by Skybloomerskullfire** [Medard22/Dolphin-MMJR2-VBI#879] [Ideas]
Button to process collected shares, in setting menu,
Sometimes it processes them and sometimes it doesn't, it is very irregular

**Comment by Medard22** [comment]
English please. Use Google translate if you have to.

**Reply by Medard22** [reply]
Sorry I have no idea what you mean. It's probably lost in translation. Please provide screenshots if possible or elaborate further.

**Reply by Skybloomerskullfire** [reply]
Sometimes it doesn't load the shaders, Shader compilation mode, such as specialized or asynchronous and synchronous, compilation before starting emulation

**Reply by Skybloomerskullfire** [reply]
![Screenshot_20250103-233623](https://github.com/user-attachments/assets/b983f469-5581-4d13-b619-653cbce91d69)

**Reply by Skybloomerskullfire** [reply]
Yes, sorry, I didn't realize

**Reply by Skybloomerskullfire** [reply]
You're my hero, Nobody has the courage to do something like you have done.

**Comment by Medard22** [comment]
Okay, I think I understand what you mean now. The short answer is that the button is not necessary. The long answer requires some explanation.

The GameCube and Wii GPU, or Flipper, has some fixed-function components but also features a programmable TEV unit. On the original consoles, the TEV unit is configured and executed immediately. However, Dolphin has to generate a specialized shader whenever it encounters a new TEV configuration. To render it, Dolphin delays the CPU thread to let the GPU and driver compile the shader.

Many shaders compile so quickly that you won’t even notice the delay, but when it takes longer, the game visibly pauses, resulting in a shader compilation stutter. To minimize this stuttering, Dolphin uses two separate shader caches: a UID cache and a pipeline cache. Both caches have a version identifier, similar to saved states.

When you start a game for the first time, the TEV configurations encountered are collected, processed, compiled, and stored in the UID and pipeline caches. This process often causes noticeable shader compilation stutters, depending on the game and your device. The next time you run the game, Dolphin checks if the shaders are already stored in the cache and uses the precompiled versions.

Here’s the key difference between the two caches:
- The UID cache is device-agnostic. It’s essentially a list of TEV configurations used to generate specialized shaders, which are then compiled. However, this cache can be often invalidated when developers make changes to the GPU pipeline. 
- The pipeline cache contains precompiled shaders, making it device AND version-specific. If you update Dolphin (which I do often when testing my CI builds), the pipeline cache is invalidated, and the shaders must be generated and compiled again.

So, what does the “Compile Shaders Before Starting” option do?
- When enabled, Dolphin checks if the pipeline cache is invalidated. If it is, it uses the UID cache to generate specialized shaders and compile them before the game starts, reducing the shader compilation stutter (similar to how modern Unreal Engine games handle shaders). If the option is disabled, the game starts immediately, but shaders are compiled on the fly during gameplay, potentially causing stutters until they’re cached.

It’s worth noting that enabling this option doesn’t guarantee the complete elimination of shader stutter, because not all TEV configurations may have been cached in the UID cache. However, the more you play the game, the smoother it becomes over time.

Why doesn’t Dolphin precompile shaders every time?
- It doesn’t need to. Once shaders are compiled and stored in the caches for the same system and Dolphin version, the emulator simply checks the cache and skips the compilation step.

What about Ubershaders?
- Ubershaders reconstruct the entire GPU pipeline into a massive set of shaders that run on the GPU. Once compiled, ubershaders eliminate the need to compile specialized shaders. Exclusive ubershaders use this approach.
- Hybrid ubershaders use ubershaders temporarily until specialized shaders are generated and compiled. Once that happens, rendering of the objects is passed to the compiled specialized shaders.

Why not use ubershaders all the time?
Ubershaders are significantly more resource-intensive and tend to run slower than specialized shaders. They are also highly complex, and GPUs and GPU drivers—especially on mobile devices—are often finicky. Using ubershaders on mobile devices often leads to crashes, graphic bugs, or poor performance. For mobile devices, specialized shaders are the way to go.

I hope this explanation makes things a bit clearer. It’s tricky to explain it without getting too technical. If you have any more questions feel free to ask.
